src/rules/no_unused_vars.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright 2020 the Deno authors. All rights reserved. MIT license.
+use super::Context;
+use super::LintRule;
+use swc_ecmascript::ast::{Expr, Ident, MemberExpr, Pat, VarDeclarator};
+use swc_ecmascript::utils::{find_ids, ident::IdentLike, Id};
+use swc_ecmascript::visit::Node;
+use swc_ecmascript::visit::Visit;
+use swc_ecmascript::visit::VisitWith;
+
+use std::collections::HashSet;
+use std::sync::Arc;
+
+pub struct NoUnusedVars;
+
+impl LintRule for NoUnusedVars {
+  fn new() -> Box<Self> {
+    Box::new(NoUnusedVars)
+  }
+
+  fn code(&self) -> &'static str {
+    "no-unused-vars"
+  }
+
+  fn lint_module(
+    &self,
+    context: Arc<Context>,
+    module: &swc_ecmascript::ast::Module,
+  ) {
+    let mut collector = Collector {
+      used_vars: Default::default(),
+    };
+    module.visit_with(module, &mut collector);
+
+    let mut visitor = NoUnusedVarVisitor::new(context, collector.used_vars);
+    module.visit_with(module, &mut visitor);
+  }
+}
+
+/// Collects information about variable usages.
+struct Collector {
+  used_vars: HashSet<Id>,
+}
+
+impl Visit for Collector {
+  // TODO(kdy1): swc_ecmascript::visit::noop_visit_type!() after updating swc
+  // It will make binary much smaller. In case of swc, binary size is reduced to 18mb from 29mb.
+  // noop_visit_type!();
+
+  fn visit_expr(&mut self, expr: &Expr, _: &dyn Node) {
+    match expr {
+      Expr::Ident(i) => {
+        // Mark the variable as used.
+        self.used_vars.insert(i.to_id());
+      }
+      _ => expr.visit_children_with(self),
+    }
+  }
+
+  fn visit_pat(&mut self, pat: &Pat, _: &dyn Node) {
+    match pat {
+      // Ignore patterns
+      Pat::Ident(..) | Pat::Invalid(..) => return,
+      //
+      _ => pat.visit_children_with(self),
+    }
+  }
+
+  fn visit_member_expr(&mut self, member_expr: &MemberExpr, _: &dyn Node) {
+    member_expr.obj.visit_with(member_expr, self);
+    if member_expr.computed {
+      member_expr.prop.visit_with(member_expr, self);
+    }
+  }
+}
+
+struct NoUnusedVarVisitor {
+  context: Arc<Context>,
+  used_vars: HashSet<Id>,
+}
+
+impl NoUnusedVarVisitor {
+  fn new(context: Arc<Context>, used_vars: HashSet<Id>) -> Self {
+    Self { context, used_vars }
+  }
+}
+
+/// As we only care about variables, only variable declrations are checked.
+impl Visit for NoUnusedVarVisitor {
+  // TODO(kdy1): swc_ecmascript::visit::noop_visit_type!() after updating swc
```

Comment:
This trick can be applied to other visitors, too.

Reply:
Sounds good 👍  `deno_lint` already uses `swc_ecmascript 0.3.2` so it can be used already?

Reply:
I'll do this.
I think doing so for all passes, which is a good fit for separate pr, will reduces binary size a lot. Rustc cannot deduce that a visitor does not interact with type.

Reply:
Sounds good 👍  `deno_lint` is actually only 650kB when compiled into `deno` but any size reduction is more than welcome